### PR TITLE
Mobilenet gpu implementation

### PR DESCRIPTION
**Pull request by NHZlX**
>The normal group convolution for mobilenet is very slow, particularly in training process. So,  i firstly implement the gpu acceleration for depthwise convolution for mobilenet 1.0 with input 224 *224. 

### current Implementation effect
|category|  batch 1  forwardbackward (s) |batch 40 forwardbackward(s)|
|----| ------------- |:-------------:|
|group convolution |0.75|29.23|
|cudnn convolution|0.74|28.88 | 
|depthwise gpu acceleration|0.052|1.27|

### future
- there is still acceleration room for gpu.
- a new pr for cpu acceleration of mobilenet.

### CODE REVIEW

File: paddle/gserver/layers/DepthwiseConvLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+#include "DepthwiseConvLayer.h"
+#include "paddle/utils/Logging.h"
+#include "paddle/utils/Stat.h"
+#include <iostream>
+
+namespace paddle {
+
+/*
+ * The calculation of the exconvt(convolution transpose (deconv) operation)
```
Comment (hedaoyuan) [inline paddle/gserver/layers/DepthwiseConvLayer.cpp]: What does this comment mean?

File: paddle/gserver/layers/DepthwiseConvLayer.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * @brief A subclass of convolution layer.
+ * This layer expands input and use matrix multiplication to
+ * calculate convolution operation.
+ *
+ * The config file api is img_conv_layer.
+ */
+
+class DepthwiseConvLayer : public ExpandConvBaseLayer {
```
Comment (hedaoyuan) [inline paddle/gserver/layers/DepthwiseConvLayer.h]: I think you can use DepthwiseConv Function in the forward and backward calculations of `ExpandConvLayer` '.

**Comment by hedaoyuan** [comment]
Check the reason why Travis CI does not pass and fix.

File: paddle/gserver/layers/DepthwiseConvLayer.h (hunk 2 above)
Comment (NHZlX) [inline paddle/gserver/layers/DepthwiseConvLayer.h]: Although they have a lot of code duplication, but there are still subtle differences， eg, the weight_multiplier_ in   depthwiseConvLayer

Comment (hedaoyuan) [inline paddle/gserver/layers/DepthwiseConvLayer.h]: Maybe put the `weight_multiplier_` in `DepthwiseConvGradFilterFunction` is better. The other two functions do not need weight_multiplier_.

File: paddle/function/DepthwiseConvOpGpu.cu [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "ConvOp.h"
+#include "DepthwiseConvOp.h"
+#include "GemmFunctor.h"
+#include "paddle/math/MemoryHandle.h"
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOpGpu.cu]: Done

File: paddle/function/DepthwiseConvOpGpu.cu [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+class DepthwiseConvFunctor<DEVICE_TYPE_GPU, T>{
+public:
+  void operator()(int outputSize, 
+            const T* inputData, 
+            const T* filterData,
+            int batchSize,
+            int outputChannels,
+            int outputHeight,
+            int outputWidth,
+			int inputHeight,
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOpGpu.cu]: Done.

File: paddle/function/DepthwiseConvOpGpu.cu [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+        //gemv
+    }
+};
+
+#ifdef PADDLE_TYPE_DOUBLE
+using real=double;
+#else 
+using real=float;
+#endif
+template class DepthwiseConvGradInputFunctor<DEVICE_TYPE_GPU, real>;
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOpGpu.cu]: Done.

File: paddle/function/DepthwiseConvOp.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include "ConvOp.h"
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.h]: Done.

File: paddle/function/DepthwiseConvOp.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+namespace paddle {
+
+template <DeviceType Device, class T>
+class DepthwiseConvFunctor {
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.h]: Done.

File: paddle/function/DepthwiseConvOp.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+                  int filterWidth,
+                  int strideH,
+                  int strideW,
+                  int paddingH,
+                  int paddingW,
+                  T* outputData);
+};
+
+template <DeviceType Device, class T>
+class DepthwiseConvGradInputFunctor {
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.h]: Done.

File: paddle/function/DepthwiseConvOp.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                  int filterWidth,
+                  int strideH,
+                  int strideW,
+                  int paddingH,
+                  int paddingW,
+                  T* inputGrad);
+};
+
+template <DeviceType Device, class T>
+class DepthwiseConvGradFilterFunctor {
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.h]: Done.

File: paddle/function/DepthwiseConvOp.cpp [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+    size_t outputHeight = output[2];
+    size_t outputWidth = output[3];
+
+    real* inputData = inputs[0].data<real>();
+    real* filterData = inputs[1].data<real>();
+    real* outputData = outputs[0].data<real>();
+    size_t outputSize = batchSize * outputChannels * outputHeight * outputWidth;
+
+    DepthwiseConvFunctor<Device, real> depthwiseConv;
+    depthwiseConv(outputSize,
```
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.cpp]: Done.

File: paddle/gserver/layers/DepthwiseConvLayer.cpp (hunk 1 above)
Comment (NHZlX) [inline paddle/gserver/layers/DepthwiseConvLayer.cpp]: I have deleted the wrong comment.

File: paddle/gserver/layers/DepthwiseConvLayer.h (hunk 2 above)
Comment (NHZlX) [inline paddle/gserver/layers/DepthwiseConvLayer.h]: the weight_multiplier_ related operations ware replaced by operation sumRows(...)  in  paddle/math/BaseMatrix.h

**Comment by NHZlX** [comment]
The clang-format version in my workspace is 5.0, so it will fail to pass the travis-ci . I will settle it ASAP.

File: paddle/function/DepthwiseConvOpTest.cpp [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+};
+
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: Add some TEST,  that the function1 is `GemmConv` and function2 is `DepthwiseConv`.

File: paddle/function/DepthwiseConvOpTest.cpp [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+#ifndef PADDLE_ONLY_CPU
+TEST(Forward, GEMM2) {
+  DepthwiseConvolutionTest<DEVICE_TYPE_GPU, DEVICE_TYPE_GPU> test(
+      "DepthwiseConv-GPU", "DepthwiseConv-GPU", kForwardTest);
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: "DepthwiseConv-GPU", "DepthwiseConv-GPU" -> "DepthwiseConv-CPU", "DepthwiseConv-GPU"

File: paddle/function/DepthwiseConvOpTest.cpp [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+  DepthwiseConvolutionTest2<DEVICE_TYPE_CPU, DEVICE_TYPE_GPU> test2(
+      "DepthwiseConv-GPU", "DepthwiseConv-GPU", kForwardTest);
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: same as above

File: paddle/function/DepthwiseConvOpTest.cpp [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+TEST(BackwardInput, GEMM) {
+  DepthwiseConvolutionTest<DEVICE_TYPE_CPU, DEVICE_TYPE_GPU> test(
+      "DepthwiseConvGradInput-GPU",
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: same as above

File: paddle/function/DepthwiseConvOpTest.cpp [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+      kBackwardInputTest);
+  DepthwiseConvolutionTest2<DEVICE_TYPE_CPU, DEVICE_TYPE_GPU> test2(
+      "DepthwiseConvGradInput-GPU",
+      "DepthwiseConvGradInput-GPU",
+      kBackwardInputTest);
+}
+
+TEST(BackwardFilter, GEMM) {
+  DepthwiseConvolutionTest<DEVICE_TYPE_CPU, DEVICE_TYPE_GPU> test(
+      "DepthwiseConvGradFilter-GPU",
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: same as above

File: paddle/function/DepthwiseConvOpTest.cpp (hunk 13 above)
Comment (NHZlX) [inline paddle/function/DepthwiseConvOpTest.cpp]: I known this, but the situation now is the cpu version depthwise conv has not been implemented.

Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpTest.cpp]: Will you add the CPU code in this pr? If not, I think it is not necessary to add these tests now. Add a TODO, explain that when the CPU version is implemented need to increase the test case.

Comment (NHZlX) [inline paddle/function/DepthwiseConvOpTest.cpp]: There will be a new pr for the cpu code.  I will add a TODO .

File: paddle/function/ConvOpTest.cpp [hunk 16]
```diff
@@ -177,6 +177,156 @@ class ConvolutionTest2 {
...
+public:
+  DepthwiseConvolutionTest(const std::string& conv1,
+                           const std::string& conv2,
+                           TestType type,
+                           std::string algo = "auto") {
+    for (size_t batchSize : {1, 32}) {
+      for (size_t inputSize : {7, 14, 54}) {
+        for (size_t filterSize : {1, 3, 5}) {
+          for (size_t inputChannels : {64, 128}) {
+            size_t outputChannels = inputChannels;
```
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: outputChannels != inputChannels的情况呢？

File: paddle/function/DepthwiseConvOpGpu.cu [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+
+        size_t blocks = (colDataSize + 1024 -1) / 1024;
+        size_t blockX = 512;
+        size_t blockY = (blocks+512-1)/512;
+        dim3 threads(1024, 1);
+        dim3 grid(blockX, blockY);
+		BaseMatrix filterGradMatrix(inputChannels * filterHeight * filterWidth, 1, filterGrad, false, true);
+
+        for(int i = 0; i < batchSize; i++) {
+			ConvolutionDepthwiseFilterBackward<T>
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpGpu.cu]: 这些代码整理一下，至少要对齐。

File: paddle/function/DepthwiseConvOpGpu.cu [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+            int inputHeight,
+            int inputWidth,
+            int filterHeight,
+            int filterWidth,
+            int strideH,
+            int strideW,
+            int paddingH,
+            int paddingW,
+            T* inputGrad){
+
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOpGpu.cu]: format可能没有检查出来，手动自己看一下，这些代码格式。有的多行空格，有的对齐比较乱。

File: paddle/gserver/layers/DepthwiseConvLayer.h (hunk 2 above)
Comment (hedaoyuan) [inline paddle/gserver/layers/DepthwiseConvLayer.h]: 这个有必要实现成一个Layer吗？

File: paddle/function/ConvOpTest.cpp [hunk 19]
```diff
@@ -228,24 +231,76 @@ TEST(Forward, GEMM) {
 #ifndef PADDLE_ONLY_CPU
 TEST(Forward, GEMM2) {
   ConvolutionTest<DEVICE_TYPE_CPU, DEVICE_TYPE_GPU> test(
-      "GemmConv-CPU", "GemmConv-GPU", kForwardTest);
+      "GemmConv-CPU", "GemmConv-GPU", kForwardTest, false);
```
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: 可以在接口参数上直接defualt=false，这里少写个参数。

File: paddle/function/DepthwiseConvOp.cpp [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+  }
+
+  void check(const BufferArgs& inputs, const BufferArgs& outputs) override {
+    const TensorShape& output = inputs[0].shape();
+    const TensorShape& input = inputs[1].shape();
+    const TensorShape& filter = outputs[0].shape();
+    checkShape(input, filter, output);
+  }
+
+  void calc(const BufferArgs& inputs, const BufferArgs& outputs) override {
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOp.cpp]: 这里面需要增加`CHECK_EQ(outputs[0].getArgType(), ADD_TO);`

File: paddle/function/DepthwiseConvOp.cpp [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+  }
+
+  void check(const BufferArgs& inputs, const BufferArgs& outputs) override {
+    const TensorShape& input = inputs[0].shape();
+    const TensorShape& filter = inputs[1].shape();
+    const TensorShape& output = outputs[0].shape();
+    checkShape(input, filter, output);
+  }
+
+  void calc(const BufferArgs& inputs, const BufferArgs& outputs) override {
```
Comment (hedaoyuan) [inline paddle/function/DepthwiseConvOp.cpp]: 需要增加`CHECK_EQ(outputs[0].getArgType(), ADD_TO);`

File: paddle/function/ConvOpTest.cpp (hunk 19 above)
Comment (NHZlX) [inline paddle/function/ConvOpTest.cpp]: 我默认的是true， 所以depthwise conv 不用传参数

File: paddle/function/DepthwiseConvOp.cpp (hunk 21 above)
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.cpp]: Done

File: paddle/function/DepthwiseConvOp.cpp (hunk 20 above)
Comment (NHZlX) [inline paddle/function/DepthwiseConvOp.cpp]: Done

File: paddle/function/ConvOpTest.cpp (hunk 19 above)
Comment (hedaoyuan) [inline paddle/function/ConvOpTest.cpp]: 也可以，不过depthwise是优化特例，所以defualt=false更好。

**Pull request merged**